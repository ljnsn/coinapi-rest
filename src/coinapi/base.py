"""Base class for operation collections."""

import enum
from typing import TypeVar

import httpx
import msgspec
from httpx import codes

from coinapi import utils
from coinapi._hooks import HookContext
from coinapi.config import CoinAPIConfig
from coinapi.models import errors
from coinapi.models.operations.base import CoinAPIRequest, CoinAPIResponse


class AcceptEnum(str, enum.Enum):
    """Valid accept header values."""

    APPLICATION_JSON = "application/json"
    TEXT_JSON = "text/json"
    TEXT_PLAIN = "text/plain"
    APPLICATION_X_MSGPACK = "application/x-msgpack"
    ANY = "*/*"


RequestT = TypeVar("RequestT", bound=CoinAPIRequest)
ResponseT = TypeVar("ResponseT", bound=CoinAPIResponse)


class Base:
    """Base class for operation collections."""

    def __init__(self, sdk_config: CoinAPIConfig) -> None:
        self.sdk_configuration = sdk_config

    def _make_request(  # noqa: PLR0912, C901
        self,
        operation_id: str,
        request: RequestT,
        response_cls: type[ResponseT],
        accept_header_override: AcceptEnum | None = None,
    ) -> ResponseT:
        """Send a request."""
        hook_ctx = HookContext(
            operation_id=operation_id,
            oauth2_scopes=[],
            security_source=self.sdk_configuration.security,
        )
        base_url = utils.template_url(*self.sdk_configuration.get_server_details())
        url = utils.generate_url(type(request), base_url, request.endpoint, request)  # type: ignore[arg-type]
        headers = {}
        data, form = None, None
        if request.method in {"POST", "PUT", "PATCH"}:
            req_content_type, data, form = utils.serialize_request_body(
                request,
                "body",
            )
            if req_content_type is not None and req_content_type not in (
                "multipart/form-data",
                "multipart/mixed",
            ):
                headers["content-type"] = req_content_type
        query_params = utils.get_query_params(type(request), request) or None  # type: ignore[arg-type]
        if accept_header_override is not None:
            headers["Accept"] = accept_header_override.value
        else:
            headers["Accept"] = (
                "application/json;q=1, text/json;q=0.8, text/plain;q=0.5, application/x-msgpack;q=0"
            )
        headers["user-agent"] = self.sdk_configuration.user_agent

        security = (
            self.sdk_configuration.security()
            if callable(self.sdk_configuration.security)
            else self.sdk_configuration.security
        )
        client = utils.configure_security_client(
            self.sdk_configuration.client,
            security,
        )

        try:
            req = self.sdk_configuration.get_hooks().before_request(
                hook_ctx,  # type: ignore[arg-type]
                httpx.Request(
                    request.method,
                    url,
                    params=query_params,
                    data=data,
                    files=form,
                    headers=headers,
                ),
            )
            http_res = client.send(req)
        except Exception as e:  # noqa: BLE001
            _, exc = self.sdk_configuration.get_hooks().after_error(hook_ctx, None, e)  # type: ignore[arg-type]
            raise exc from e  # type: ignore[misc]

        if utils.match_status_codes(["4XX", "5XX"], http_res.status_code):
            http_res, exc = self.sdk_configuration.get_hooks().after_error(  # type: ignore[assignment]
                hook_ctx,  # type: ignore[arg-type]
                http_res,
                None,
            )
            if exc:
                raise exc
        else:
            result = self.sdk_configuration.get_hooks().after_success(
                hook_ctx,  # type: ignore[arg-type]
                http_res,
            )
            if isinstance(result, Exception):
                raise result
            http_res = result

        content_type = http_res.headers.get("Content-Type", "")

        res = response_cls(
            status_code=http_res.status_code,
            content_type=content_type,
            raw_response=http_res,
        )

        if httpx.codes.is_success(http_res.status_code):
            if utils.match_content_type(content_type, "text/plain"):
                res.content_plain = http_res.text
            elif utils.match_content_type(
                content_type,
                "application/json",
            ) or utils.match_content_type(content_type, "text/json"):
                content_cls = next(
                    field
                    for field in msgspec.structs.fields(response_cls)
                    if field.name == "content"
                ).type
                out = msgspec.json.decode(http_res.content, type=content_cls)
                res.content = out
            elif utils.match_content_type(content_type, "application/x-msgpack"):
                res.body = http_res.content
            else:
                msg = f"unknown content-type received: {content_type}"
                raise errors.CoinAPIError(
                    msg,
                    http_res.status_code,
                    http_res.text,
                    http_res,
                )
        elif codes.is_client_error(http_res.status_code) or codes.is_server_error(
            http_res.status_code,
        ):
            raise errors.CoinAPIError(
                "API error occurred",
                http_res.status_code,
                http_res.text,
                http_res,
            )

        return res
