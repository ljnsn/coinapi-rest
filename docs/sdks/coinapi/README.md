# CoinAPI SDK


## Overview

REST API: 
RESTful endpoint provides the widest range of data, based on HTTP protocol which works in Request-Reply scheme.

Implemented Standards:

 * [HTTP1.0](https://datatracker.ietf.org/doc/html/rfc1945)
 * [HTTP1.1](https://datatracker.ietf.org/doc/html/rfc2616)
 * [HTTP2.0](https://datatracker.ietf.org/doc/html/rfc7540)
 * [OpenAPI v3](https://www.openapis.org/)

> **Note:** We adhere to the OpenAPI standards for documenting our API.

## OpenAPI Specification

To access our API's OpenAPI specification, you can use the following link: [OpenAPI v3](https://raw.githubusercontent.com/coinapi/coinapi-sdk/master/data-api/coinapi-marketdata-rest.yaml)

If you need to import the OpenAPI file into software like Postman, simply copy and paste the link below:
```shell
https://raw.githubusercontent.com/coinapi/coinapi-sdk/master/data-api/coinapi-marketdata-rest.yaml
```

## Endpoints

Enviroment | Encryption | Value 
--- | --- | ---
Production | Yes | `https://rest.coinapi.io/`
Production | No | `http://rest.coinapi.io/`

:::info

For real-time market data streaming, you should use WebSockets. REST API only supports pooling, meaning you can periodically request the current market data state. In streaming, you subscribe and data or updates are delivered to you continuously.

:::

## General

If you want to learn how to authenticate to this API, you can find detailed instructions and guidance in
[authentication section](/authentication) of this documentation.

### HTTP Requests

Each HTTP request must contain the header ``Accept: application/json`` as all our responses are in JSON format.

We encourage you to use the HTTP request header ``Accept-Encoding: deflate, gzip`` for all requests. 
This will indicate to us that we can deliver compressed data to you which on your side should be decompressed transparently.

:::tip

By allowing data compression you are lowering bandwidth requirements by approximately 80%. 
This is important for requesting large amounts of data or using WebSocket Streaming API, 
as we can deliver data to you faster and more effectively.

:::

#### HTTP Success

Successful HTTP responses have the status code `200` and the body in a format according to documentation of the requested resource.

:::info

You should always check that your HTTP response status code is equal to 200, otherwise the requested was not successful.

:::

#### HTTP Errors

> Error message is returned in JSON structured like this:

```json
{
    "error": "Invalid API key"
}
```

All HTTP requests with response status codes different to `200` must be considered as failed 
and you should expect additional JSON inside the body of the response with the error message encapsulated inside it as shown in the example.
We use the following error codes:

Error Code | Meaning
---------- | -------
400 | Bad Request -- There is something wrong with your request
401 | Unauthorized -- Your API key is wrong
403 | Forbidden -- Your API key doesnt't have enough privileges to access this resource
429 | Too many requests -- You have exceeded your API key rate limits
550 | No data -- You requested specific single item that we don't have at this moment.

:::info

Good practice is to store all error messages somewhere along with request data for further manual review.

:::

### Limits

Any authenticated endpoint is providing (in HTTP response headers) information about the current state of the limits associated with API Key. In this section we will describe each limit.

#### Request limit / APIKey

```html
X-RateLimit-Limit: 1000000
X-RateLimit-Remaining: 999989
X-RateLimit-Request-Cost: 1
```

The request limit define number of maximum requests that could be executed in the 24 hours period (sliding/rollowing window - always last 24 hours from specific moment) for your subscription.

We define request as data request credits and this is not always equal to the number of API calls executed against the API. A request is deemed to be a single one if the limit query parameter on the endpoint isn’t available, isn’t used or it's stated otherwise in the API documentation. Otherwise — if the limit query parameter is available and is used — then each of the 100 data points returned in the response is counted as one request.

For example at the 2019-08-22 13:00 UTC value of the requests remaining (X-RateLimit-Remaining) will be equal to the allocated quota (X-RateLimit-Limit) decreased by the sum of the request costs (SUM(X-RateLimit-Request-Cost)) executed in the period 2019-08-21 13:00 UTC - 2019-08-22 13:00 UTC (last 24 hours).


HTTP Header | Type | Description
---------- | ------- | ---
X-RateLimit-Used | int | Provides information about the request limit that has been used within the last 24-hour period. This header indicates the amount of request capacity consumed based on the usage history. It is important to note that the header is not always appended to every request to optimize the operation of the API.
X-RateLimit-Limit | int | Is an optional feature that can be enabled via the customer portal to impose a limit on the capabilities of a specific API key. It allows you to define a threshold for the number of requests that can be made using a single API key within a 24-hour time frame.
X-RateLimit-Remaining | int | Provides information about the number of requests that can still be made within the last 24-hour period based on the usage history. This header serves as a helpful indicator of the remaining request capacity, allowing API consumers to manage their usage effectively. It is important to note that the header is not always appended to every request to optimize the operation of the API.
X-RateLimit-Request-Cost | int | The number of requests used to generate current HTTP response.
X-RateLimit-Quota-Overage | string | Provides information about whether a given API key may exceed the plan quota within a 24-hour time frame, which could result in additional charges. This header is fully defined and configured in the customer portal.
X-RateLimit-Quota-Allocated | string |  Total number of requests that can be made within a specific subscription during a 24-hour time frame. This quota allocation is determined based on the user's subscription purchase.
X-RateLimit-Quota-Remaining | string | Provides valuable information about the remaining quota within the subscription for making requests within a 24-hour time frame. This header indicates the number of requests that can still be made within the allocated quota for the current 24-hour period.

```json
GET v1/exchanges/ECB/apiKey-ED802AF4-E855-YOUR-API-KEY
Host: coinapi.io
X-RateLimit-Used: 1000
X-RateLimit-Limit: 5000
X-RateLimit-Remaining: 4000
X-RateLimit-Request-Cost: 1
X-RateLimit-Quota-Overage: ENABLED
X-RateLimit-Quota-Allocated: 10000
X-RateLimit-Quota-Remaining: 5000
```

Explanation:

- X-RateLimit-Used: 1000 (requests used in the last 24 hours)
- X-RateLimit-Limit: 5000 (total request limit within a 24-hour time frame)
- X-RateLimit-Remaining: 4000 (requests remaining within the last 24 hours)
- X-RateLimit-Request-Cost: 1 (cost or "weight" of each individual request)
- X-RateLimit-Overage: ENABLED (API key may exceed the plan quota within a 24-hour time frame)
- X-RateLimit-Quota-Allocated: 10000 (total number of requests allowed for all API keys within the subscription within a 24-hour time frame)
- X-RateLimit-Quota-Remaining: 5000 (requests remaining within the subscription's allocated quota within the last 24 hours)

#### Concurrency limit / APIKey

```html
X-ConcurrencyLimit-Limit: 10
X-ConcurrencyLimit-Remaining: 5
```

The concurrency limit defines the number of maximum concurrent API calls/requests that the API could process for your subscription at the current moment. Every API call/request increases the Concurrency limit against quota, and when it finishes, decreases it.

HTTP Header | Type | Description
---------- | ------- | ---
X-ConcurrencyLimit-Limit | int | Concurrency limit allocated for your API key.
X-ConcurrencyLimit-Remaining | int | The number of concurrent API calls/requests available to be executed in this moment for your API key.

### Output data format

By default we are using JSON output data format for all of our endpoints, you can control format of data by using `output_format` variable in query string parameters.

#### URL Parameters

Parameter | Type | Description
---------- | ------- | -------
output_format | string | Output data format *(optional, default value is `json`, possible values are `json`, `xml` or `csv`)*
csv_include_header | bool | Ignore header line in CSV output? *(optional, default value is `true`, `true` to include CSV header line, `false` otherwise)*
csv_include_quotes | bool | Encapsulate strings with quotes in CSV output? *(optional, default value is `false`, `true` to encapsulate all strings with `"`, `false` to leave them unquoted)*
csv_exclude_col | string | Comma delimited list of column names to ignore in CSV output *(optional, by default all columns are included)*
csv_set_delimiter | string | Character that will be used as column delimiter in CSV output *(optional, default value is `;`)*
csv_set_dec_mark | string | Character that will be used as decimal separator in CSV output *(optional, default value is `.`)*
csv_set_timeformat | string | Format for datetime type in CSV output or `unix` for unix timestamp *(optional, default value is `yyyy-MM-ddTHH:mm:ss.fffffffZ`)*
csv_set_newline | string | New line type *(optional, default value is `unix`, possible values `win`, `mac`, `unix`)*

### Excel / G-Sheets

There are several ways to use data from our REST API inside the Excel, Google Sheets, or similar calculation sheet application. This section will do as best as possible to keep all information up to date on how you could load the data into these applications. Feel free to contact support if we are missing an option.

#### CSV download, import:

 1. Open the data in the CSV format from the browser eg. ```https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv```
 2. Save the data to the file with the .csv extension.
 3. Use the file saved and import it into the software.
 4. When configuring import, refer to the parameters like delimiter from the [Output data format](#output-data-format)

The platform-independent way described above is based on CSV but could also be used in other formats like JSON and XML as long as the software support it, but the import procedure needs to be adjusted accordingly.

#### Microsoft Excel

 * Use [PowerQuery](https://docs.microsoft.com/en-us/power-query/power-query-what-is-power-query) to load the URL directly into the CSV import without saving the file locally. 
 * Use the [=WEBSERVICE](https://support.office.com/en-us/article/webservice-function-0546a35a-ecc6-4739-aed7-c0b7ce1562c4) function to load the API response directly into the sheet, but this will not parse the data; additional processing is required.

#### Google Sheets

 * Use [=IMPORT](https://support.google.com/docs/answer/3093335?hl=en) function to load the REST API endpoint and automatically parse the CSV format data into the cells. eg. ```=IMPORTDATA("https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv```

#### OpenOffice Calc

 * Select the menu Insert -> Sheet From File, 2. In the Insert dialog, put the URL eg. ```https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv``` in the File Name box at the bottom. Set the drop-down list next to that to Web Page Query and click Open. The Text Import dialog opens where you can change the defaults if needed.
                    

### Available Operations

