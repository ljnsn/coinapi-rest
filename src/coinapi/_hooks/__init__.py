"""Hooks."""

from coinapi._hooks.hook_types import *
from coinapi._hooks.hooks import *
