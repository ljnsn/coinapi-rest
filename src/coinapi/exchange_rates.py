"""Exchange rates module."""

from coinapi.base import AcceptEnum, Base
from coinapi.models import operations


class ExchangeRates(Base):
    r"""Exchange rate is defined as (VWAP-24H) last 24 hour (rolling window over time) Volume Weighted Average Price across multiple data sources listed on our platform. We are selecting and managing the data sources that are used in the calculation based on multiple factors to provide data of highest quality.

    Algorithm is described below:

      1. Exchange rates are produced from quotes, trades, and metadata datasets.
      1. Symbols that are not data_type = \"SPOT\" are excluded from the calculation.
      1. Symbols from the data sources that were marked by us as not legitimate are excluded from the calculation.
      1. Quotes data where the spread is outside the range of ```<0$; 67%>``` are discarded. `spreadPrc = (ask - bid) / ((ask + bid) / 2)`
      1. The midpoint from the quote data is used as a pricing reference and it's weighted by the passive cumulative volume resting on the best prices.
      1. Volume from the trades is used to weight the midpoint prices in the VWAP24 algorithm.
      1. Midpoint data that has not been updated in the last 5 minutes and 1 second is discarded.
      1. The last 24-hour volume for each symbol is updated every 4 hours when approximately 20% of the data in the sliding window changes (also, the list of eligible markets is updated at the same time).
      1. Everywhere in the algorithm below, we are using asset pairs only from exchanges that have the highest legitimacy rank, and the rest of the exchanges are discarded. As we establish the highest-ranking exchanges that have this data for each asset pair, we ensure that the highest quality data is used for each of them. The rank used for asset pairing is carried over to the following steps.
      1. Every 1 second, we update VWAP24 data for every asset pair across all data sources.
      1. For each asset pair, we also discard data that is outside the 3 sigma range if there are at least 3 exchanges for this asset pair.
      1. From the VWAP24 data, we are creating a tree structure where node/vertex = asset and edge = rate.
      1. By traversing the tree structure using the BFS algorithm and our secret sauce, we are able to establish the final exchange rates.
    """

    def get_v1_specific_rate(
        self,
        asset_id_base: str,
        asset_id_quote: str,
        time: str | None = None,
        accept_header_override: AcceptEnum | None = None,
    ) -> operations.GetV1SpecificRateResponse:
        r"""[exchange rates] Get specific rate.

        Retrieves the exchange rate for a specific base and quote asset at a given time or the current rate.

        :::info
        If you are using an exchange rate for mission-critical operations, then for best reliability, you should measure the difference between current time and the time returned from the response to ensure that value of the difference between those meets your internal requirements.
        :::
        """
        return self._make_request(
            "Get specific rate",
            operations.GetV1SpecificRateRequest(
                asset_id_base=asset_id_base,
                asset_id_quote=asset_id_quote,
                time=time,
            ),
            operations.GetV1SpecificRateResponse,
            accept_header_override=accept_header_override,
        )

    def get_v1_base_rates(  # noqa: PLR0913
        self,
        asset_id_base: str,
        filter_asset_id: str | None = None,
        invert: bool | None = None,
        time: str | None = None,
        accept_header_override: AcceptEnum | None = None,
    ) -> operations.GetV1BaseRatesResponse:
        r"""[exchange rates] Get all current rates.

        Get the current exchange rate between requested asset and all other assets.

        :::info
        If you are using an exchange rate for mission-critical operations, then for best reliability, you should measure the difference between current time and the time returned from the response to ensure that value of the difference between those meets your internal requirements.
        :::

        :::info
        You can invert the rates by using Y = 1 / X equation, for example BTC/USD = 1 / (USD/BTC);
        :::
        """
        return self._make_request(
            "get_/v1/exchangerate/{asset_id_base}",
            operations.GetV1BaseRatesRequest(
                asset_id_base=asset_id_base,
                filter_asset_id=filter_asset_id,
                invert=invert,
                time=time,
            ),
            operations.GetV1BaseRatesResponse,
            accept_header_override=accept_header_override,
        )

    def get_v1_history_periods(
        self,
        accept_header_override: AcceptEnum | None = None,
    ) -> operations.GetV1HistoryPeriodsResponse:
        r"""[exchange rates] Timeseries periods.

        You can also obtain historical exchange rates of any asset pair, grouped into time periods.
        Get full list of supported time periods available for requesting exchange rates historical timeseries data.

        ## Timeseries periods
        Time unit |	Period identifiers
        --- | ---
        Second | 1SEC, 2SEC, 3SEC, 4SEC, 5SEC, 6SEC, 10SEC, 15SEC, 20SEC, 30SEC
        Minute | 1MIN, 2MIN, 3MIN, 4MIN, 5MIN, 6MIN, 10MIN, 15MIN, 20MIN, 30MIN
        Hour | 1HRS, 2HRS, 3HRS, 4HRS, 6HRS, 8HRS, 12HRS
        Day | 1DAY, 2DAY, 3DAY, 5DAY, 7DAY, 10DAY
        """
        return self._make_request(
            "get_/v1/exchangerate/history/periods",
            operations.GetV1HistoryPeriodsRequest(),
            operations.GetV1HistoryPeriodsResponse,
            accept_header_override=accept_header_override,
        )

    def get_v1_pair_history(
        self,
        request: operations.GetV1PairHistoryRequest,
        accept_header_override: AcceptEnum | None = None,
    ) -> operations.GetV1PairHistoryResponse:
        r"""[exchange rates] Timeseries data.

        Get the historical exchange rates between two assets in the form of the timeseries.
        """
        return self._make_request(
            "get_/v1/exchangerate/{asset_id_base}/{asset_id_quote}/history",
            request,
            operations.GetV1PairHistoryResponse,
            accept_header_override=accept_header_override,
        )
