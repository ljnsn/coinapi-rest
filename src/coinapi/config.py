"""SDK Config."""

from collections.abc import Callable
from importlib.metadata import version
from typing import ClassVar

import httpx
import msgspec

from coinapi._hooks import SDKHooks
from coinapi.models import components
from coinapi.utils import utils

SERVERS = [
    "https://rest.coinapi.io",
]
"""Contains the list of available CoinAPI servers"""


class CoinAPIConfig(msgspec.Struct):
    """The configuration for the SDK."""

    version: ClassVar[str] = version("coinapi-rest")

    client: httpx.Client | None
    security: components.Security | Callable[[], components.Security] | None = None
    server_url: str | None = ""
    server_idx: int | None = 0
    openapi_doc_version: str = "v1"
    user_agent: str = "coinapi-rest/python 0.0.1 CoinAPI v1"
    _hooks: SDKHooks | None = None

    def get_server_details(self) -> tuple[str, dict[str, str]]:
        """Get the server details."""
        if self.server_url:
            return utils.remove_suffix(self.server_url, "/"), {}
        if self.server_idx is None:
            self.server_idx = 0

        return SERVERS[self.server_idx], {}

    def get_hooks(self) -> SDKHooks:
        """Get the hooks."""
        assert self._hooks is not None  # noqa: S101
        return self._hooks
