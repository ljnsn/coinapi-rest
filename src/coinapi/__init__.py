"""CoinAPI REST SDK."""

__all__ = ("CoinAPI", "SDKConfiguration")

from coinapi.config import SDKConfiguration
from coinapi.rest import CoinAPI
