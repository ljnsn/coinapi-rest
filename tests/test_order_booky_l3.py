"""Tests for order book L3 requests."""

import pytest
from syrupy.assertion import SnapshotAssertion

from coinapi import CoinAPI


@pytest.mark.vcr()
def test_get_v1_orderbooks3_current(
    coinapi: CoinAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test for OrderBookL3.get_v1_orderbooks3_current."""
    response = coinapi.order_book_l3.get_v1_orderbooks3_current(
        filter_symbol_id="KRAKEN_SPOT_BTC_USD,KRAKEN_SPOT_BTC_EUR",
        limit_levels=10,
    )

    assert response.status_code == 200
    assert response.content is not None
    assert response.content == snapshot


@pytest.mark.vcr()
def test_get_v1_orderbooks3_symbol_id_current(
    coinapi: CoinAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test for OrderBookL3.get_v1_orderbooks3_symbol_id_current."""
    response = coinapi.order_book_l3.get_v1_orderbooks3_symbol_id_current(
        symbol_id="KRAKEN_SPOT_BTC_USD",
        limit_levels=10,
    )

    assert response.status_code == 200
    assert response.content is not None
    assert response.content == snapshot
