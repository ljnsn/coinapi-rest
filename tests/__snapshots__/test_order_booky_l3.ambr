# serializer version: 1
# name: test_get_v1_orderbooks3_current
  list([
    V1OrderBookBase(symbol_id='KRAKEN_SPOT_BTC_USDC', time_exchange=datetime.datetime(2024, 3, 6, 20, 14, 43, 237632, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 14, 43, 188510, tzinfo=datetime.timezone.utc), asks=[{'price': 67239.0, 'size': 0.00011}, {'price': 67246.34, 'size': 0.00030386}, {'price': 67254.69, 'size': 0.37172132}, {'price': 67259.19, 'size': 0.0464}, {'price': 67260.21, 'size': 0.37169082}, {'price': 67263.8, 'size': 0.00442438}, {'price': 67264.3, 'size': 0.37166822}, {'price': 67272.65, 'size': 0.0464}, {'price': 67279.84, 'size': 0.37158238}, {'price': 67290.44, 'size': 0.02328031}], bids=[{'price': 67238.99, 'size': 1.22434424}, {'price': 67235.56, 'size': 0.5045}, {'price': 67235.03, 'size': 0.00371928}, {'price': 67231.07, 'size': 0.0464}, {'price': 67230.69, 'size': 0.00595255}, {'price': 67225.53, 'size': 0.0464}, {'price': 67221.44, 'size': 0.37190513}, {'price': 67217.23, 'size': 0.0464}, {'price': 67216.76, 'size': 0.0471621}, {'price': 67211.82, 'size': 0.178}]),
    V1OrderBookBase(symbol_id='KRAKEN_SPOT_BTC_USDT', time_exchange=datetime.datetime(2024, 3, 6, 20, 14, 43, 298579, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 14, 43, 244971, tzinfo=datetime.timezone.utc), asks=[{'price': 67142.7, 'size': 0.00078997}, {'price': 67144.5, 'size': 0.00028331}, {'price': 67145.1, 'size': 0.01507401}, {'price': 67146.9, 'size': 0.37171943}, {'price': 67149.4, 'size': 0.37169765}, {'price': 67149.6, 'size': 0.00149607}, {'price': 67150.7, 'size': 0.00261511}, {'price': 67153.1, 'size': 0.00017281}, {'price': 67155.2, 'size': 0.37166539}, {'price': 67155.9, 'size': 0.05}], bids=[{'price': 67142.2, 'size': 0.37174477}, {'price': 67138.4, 'size': 0.37176606}, {'price': 67135.4, 'size': 0.0885}, {'price': 67133.3, 'size': 0.37178624}, {'price': 67132.2, 'size': 0.056}, {'price': 67129.9, 'size': 0.000288}, {'price': 67129.0, 'size': 0.05}, {'price': 67128.9, 'size': 0.000127}, {'price': 67125.6, 'size': 0.0885}, {'price': 67124.8, 'size': 0.0885}]),
    V1OrderBookBase(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 14, 43, 309490, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 14, 43, 258857, tzinfo=datetime.timezone.utc), asks=[{'price': 67279.2, 'size': 6.95850934}, {'price': 67279.3, 'size': 0.0005}, {'price': 67280.8, 'size': 0.02134208}, {'price': 67280.9, 'size': 1.48630701}, {'price': 67281.0, 'size': 1.48630418}, {'price': 67284.0, 'size': 0.4981}, {'price': 67284.6, 'size': 0.08447625}, {'price': 67285.7, 'size': 0.05951342}, {'price': 67286.0, 'size': 0.42154749}, {'price': 67286.6, 'size': 2.22927135}], bids=[{'price': 67279.1, 'size': 1.61691018}, {'price': 67258.5, 'size': 1.48680086}, {'price': 67258.4, 'size': 0.44593287}, {'price': 67257.9, 'size': 0.44600381}, {'price': 67257.0, 'size': 0.44593287}, {'price': 67256.3, 'size': 1.4868495}, {'price': 67254.2, 'size': 1.48689435}, {'price': 67251.9, 'size': 0.375}, {'price': 67249.9, 'size': 0.0514}, {'price': 67248.5, 'size': 1.48702106}]),
    V1OrderBookBase(symbol_id='KRAKEN_SPOT_BTC_EUR', time_exchange=datetime.datetime(2024, 3, 6, 20, 14, 43, 309352, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 14, 43, 258665, tzinfo=datetime.timezone.utc), asks=[{'price': 61702.7, 'size': 0.84336354}, {'price': 61704.8, 'size': 0.001}, {'price': 61722.1, 'size': 0.2234}, {'price': 61722.2, 'size': 0.264}, {'price': 61722.4, 'size': 0.25}, {'price': 61726.0, 'size': 0.07445}, {'price': 61726.1, 'size': 0.00834674}, {'price': 61726.2, 'size': 0.2053}, {'price': 61726.3, 'size': 0.04443074}, {'price': 61729.4, 'size': 0.48538464}], bids=[{'price': 61702.6, 'size': 0.33870135}, {'price': 61702.5, 'size': 0.008}, {'price': 61701.3, 'size': 0.29733543}, {'price': 61691.9, 'size': 0.0024991}, {'price': 61691.8, 'size': 0.02300379}, {'price': 61691.7, 'size': 0.10292988}, {'price': 61688.8, 'size': 0.04443074}, {'price': 61688.6, 'size': 0.7}, {'price': 61688.5, 'size': 0.06613447}, {'price': 61688.3, 'size': 0.495}]),
  ])
# ---
# name: test_get_v1_orderbooks3_symbol_id_current
  V1OrderBookBase(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 14, 59, 689389, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 14, 59, 632791, tzinfo=datetime.timezone.utc), asks=[{'price': 67279.2, 'size': 7.44848028}, {'price': 67279.3, 'size': 0.03306523}, {'price': 67279.4, 'size': 1.48634089}, {'price': 67280.2, 'size': 1.48632268}, {'price': 67281.0, 'size': 1.48630566}, {'price': 67282.9, 'size': 0.001}, {'price': 67284.0, 'size': 0.4981}, {'price': 67284.6, 'size': 0.05036925}, {'price': 67285.4, 'size': 2.22931136}, {'price': 67285.7, 'size': 0.05951342}], bids=[{'price': 67279.1, 'size': 1.61591794}, {'price': 67275.0, 'size': 0.0524}, {'price': 67266.0, 'size': 0.0514}, {'price': 67264.7, 'size': 0.44595673}, {'price': 67263.4, 'size': 0.89186243}, {'price': 67260.9, 'size': 1.48674676}, {'price': 67259.7, 'size': 1.48677347}, {'price': 67257.5, 'size': 1.48682116}, {'price': 67254.0, 'size': 1.48689994}, {'price': 67251.9, 'size': 0.375}])
# ---
